### PR TITLE
Page mode and post mode scope adjusted

### LINKED ISSUE #74: The different modes do not have the right scope
## Problem

As for now in the frontend, the `user`, `page` and `post` mode are at the same level such as:
```clojure
;; initial db
{:app/current-view (rfe/push-state :flybot/home)
 :app/theme        local-store-theme
 :user/mode        :reader
 :page/mode        :read ;;wrong scope
 :post/mode        :read ;; wrong scope
 :nav/navbar-open? false}
```
This is conceptually wrong as each page should have its mode and each post should have its mode.

This has for consequence some overcomplexity in the logic at the post level which could be easily avoided via individual post modes and individual page modes.

## Suggestion

- Change flat pages and posts maps to indexed maps to better represent the front-end layers.
- Move `:page/mode` to page level and `:post/mode` to post level
- A new post and an existing post should have the same edit component: a new post just has a temp id.
- Obvious refactoring should result from those improvements

### PR DESCRIPTION
**Pull request by skydread1**
Closes #74
---
[Frontend]
- Each post as its `:post/mode` (:read or :edit)
- Each page as its `:page/mode` (:read or :edit)
- Pages and Posts data structure are now indexed by `:page/name` and `:post/id`
- Post logic is now the same for a new post or existing post
- Some refactoring were done thanks to the above improvements

### CODE REVIEW

File: src/cljc/flybot/validation.cljc [hunk 1]
```diff
@@ -65,16 +65,22 @@
...
-       (-> fields
-           (dissoc :post/view)
+           (dissoc :post/view :post/mode)
            (assoc :post/id (random-uuid)
-                  :post/page page-name
-                  :post/creation-date (js/Date.))))))
+                  :post/creation-date (js/Date.)))
+       (-> fields
+           (dissoc :post/view :post/mode)
+           (assoc :post/last-edit-date (js/Date.))))))
```
Comment (robertluo) [inline src/cljc/flybot/validation.cljc:78]: - should `fields` be named as `post`?
 - code duplicate, not clear

```suggestions
(let [temp-id? (temperary-id? post)
      date-field (if temp-id? :post/creation-date :post/last-edit-date)]
  (-> post
      (dissoc :post/view :post/mode)
      (update :post/id #(if temp-id? identity random-uuid))
      (assoc date-field (js/Date.))))
```

File: src/cljs/flybot/db.cljs [hunk 2]
```diff
@@ -204,16 +206,17 @@
...
-   (if (= :edit (:page/mode db))
-     {:db (assoc db :page/mode :read)}
-     {:db (assoc db :page/mode :edit)
-      :fx [[:dispatch [:evt.post/set-mode :read]]]})))
+ [(rf/path :app/pages)]
+ (fn [pages [_ page-name]]
+   (let [mode (-> pages page-name :page/mode)]
+     (if (= :edit mode)
+       (assoc-in pages [page-name :page/mode] :read)
+       (assoc-in pages [page-name :page/mode] :edit)))))
```
Comment (robertluo) [inline src/cljs/flybot/db.cljs:219]: code duplicate. Because we always `assoc-in`, suggest make this intention clearer.

**Comment by robertluo** [comment]
The result looks nice. 

Do we got cljs unit tests? Inline the effects using anonymous functions does not look really nice, and might need regression tests.

**Comment by skydread1** [comment]
> The result looks nice.
> 
> Do we got cljs unit tests? Inline the effects using anonymous functions does not look really nice, and might need regression tests.

We don't have the cljs tests yet, should I create another issue for it and send another PR?

**Review by robertluo — APPROVED** [review]

**Comment by robertluo** [comment]
>> We don't have the cljs tests yet, should I create another issue for it and send another PR? 

:+1:

**Pull request merged**